### ISSUE TITLE
Flutter Web misses rapid shortcut presses

### ISSUE THREAD

**Issue by luigi-rosso** [flutter/flutter#95474]
Tested in 2.8.0 and master.

I'm using [Shortcuts and Actions](https://docs.flutter.dev/development/ui/advanced/actions_and_shortcuts) to test triggering custom actions via keybindings. The details below are regarding a PasteAction via a PasteIntent when pressing command v, but the general problem/root cause is due to Flutter Web's [keyboarding binding](https://github.com/flutter/engine/blob/9177b32b75a8dfa3deb4f8823003ac219cea7c13/lib/web_ui/lib/src/engine/keyboard_binding.dart#L409-L412) and happens when listening to HardwareKeyboard or RawKeyboard events too.

On the Web, Flutter does this [debounce trick](https://github.com/flutter/engine/blob/9177b32b75a8dfa3deb4f8823003ac219cea7c13/lib/web_ui/lib/src/engine/keyboard_binding.dart#L312-L321) to synthesize key-ups on modifier key combos (I remember fighting this 6 years ago, can't believe browsers still don't report key-ups in these cases). In combo with [ignoring presses](https://github.com/flutter/engine/blob/9177b32b75a8dfa3deb4f8823003ac219cea7c13/lib/web_ui/lib/src/engine/keyboard_binding.dart#L409-L412) without a release, Flutter web misses reporting combos like (note these are quick, time is important as the keyguards have 1-2s durations):

0ms: Press CMD
200ms: Press V
400ms: Release V
600ms: Press V
800ms: Release CMD

Which should result in two Paste actions, instead only generates one because the release for V isn't seen (browser issue). Flutter web will report:

0ms: Press CMD
200ms: Press V
800ms: Release CMD
2200ms: Release V (Flutter's keydown guard on mac is 2s)

Our workaround takes a different approach, we will always report keydowns when modifiers are pressed and instead of synthesizing on a timer, we synthesize key-ups for all known pressed keys when the modifier is finally released. Rive's fix will report:

0ms: Press CMD
200ms: Press V
600ms: Press V (we could also synthesize a key-up right before this, but we currently don't).
800ms: Release V (sent just before CMD as we see V was pressed while CMD was down, so the release of CMD releases all other pressed keys in context of it)
800ms: Release CMD

We currently do this by spinning up our own platform keyboard channel. 

We're planning on forking the engine to do this for Rive to keep our hack out of our app, but I wanted to bring it up to see if the Flutter team has interest in us attempting to PR this back, or if others are also running into a similar issue perhaps a more general solution could be found.

**Comment by maheshmnj** [comment]
HI @luigi-rosso, Thanks for filing the issue. I am able to reproduce the issue. On rapidly pressing shortcuts like `cmd+v` on the Web takes a certain time duration(around 2000ms) between consecutive keystrokes while on the Desktop it's instant.  

_In both the videos I am constantly pressing `cmd+v`, On the left console shows the key events and the time elapsed_

### Web
https://user-images.githubusercontent.com/31410839/147089046-67d85834-c0d4-45bb-bc2c-8c47010e4423.mov

### Desktop

https://user-images.githubusercontent.com/31410839/147089805-933673dc-7fc4-4225-8c75-9c4e1364c619.mov



<details>
<summary>code sample</summary>

```dart
import 'package:flutter/material.dart';
import 'package:flutter/services.dart';

void main() {
  runApp(MyApp());
}

class MyApp extends StatefulWidget {
  MyApp({Key? key}) : super(key: key);

  @override
  State<MyApp> createState() => _MyAppState();
}

class _MyAppState extends State<MyApp> {
  Stopwatch stopwatch = Stopwatch();

  @override
  void initState() {
    // TODO: implement initState
    super.initState();
    stopwatch.start();
  }

  @override
  void dispose() {
    // TODO: implement dispose
    super.dispose();
    stopwatch.stop();
  }

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
        home: GlobalKeyboardListener(
            onKeyEvent: (KeyEvent event) {
              if (event is KeyDownEvent) {
                if (HardwareKeyboard.instance.logicalKeysPressed.isNotEmpty) {
                  final pressedKeys = LogicalKeySet.fromSet(
                    HardwareKeyboard.instance.logicalKeysPressed,
                  );
                  // ignore: avoid_print
                  print(
                      'elapsed ${stopwatch.elapsedMilliseconds}ms ${pressedKeys}');
                }
              }
              return false;
            },
            child: Scaffold(
                appBar: AppBar(
                  title: const Text('Flutter Home page'),
                ),
                body: Column(
                    mainAxisAlignment: MainAxisAlignment.center,
                    children: const [
                      TextField(),
                    ]))));
  }
}

class GlobalKeyboardListener extends StatefulWidget {
  final bool Function(KeyEvent keyEvent) onKeyEvent;
  final Widget child;

  const GlobalKeyboardListener({
    required this.onKeyEvent,
    required this.child,
    Key? key,
  }) : super(key: key);

  @override
  State<GlobalKeyboardListener> createState() {
    return _GlobalKeyboardListenerState();
  }
}

class _GlobalKeyboardListenerState<T> extends State<GlobalKeyboardListener> {
  @override
  void initState() {
    startListen();
    super.initState();
  }

  @override
  void dispose() {
    stopListen();
    super.dispose();
  }

  @override
  Widget build(BuildContext context) {
    return widget.child;
  }

  void startListen() {
    HardwareKeyboard.instance.addHandler(widget.onKeyEvent);
  }

  void stopListen() {
    HardwareKeyboard.instance.removeHandler(widget.onKeyEvent);
  }
}

```
</details>


<details>
<summary>logs(web)</summary>

```
Restarted application in 171ms.
elapsed 10397ms LogicalKeySet#4b191(keys: Meta Left)
elapsed 10765ms LogicalKeySet#10264(keys: Meta Left + Key V)
elapsed 12833ms LogicalKeySet#10264(keys: Meta Left + Key V)
elapsed 14905ms LogicalKeySet#10264(keys: Meta Left + Key V)
elapsed 16990ms LogicalKeySet#10264(keys: Meta Left + Key V)
elapsed 19104ms LogicalKeySet#10264(keys: Meta Left + Key V)
elapsed 21264ms LogicalKeySet#10264(keys: Meta Left + Key V)
elapsed 23348ms LogicalKeySet#10264(keys: Meta Left + Key V)
elapsed 25512ms LogicalKeySet#10264(keys: Meta Left + Key V)
elapsed 27731ms LogicalKeySet#10264(keys: Meta Left + Key V)
elapsed 29774ms LogicalKeySet#10264(keys: Meta Left + Key V)
Application finished.
Exited (sigterm)
```
</details>


<details>
<summary>logs (Desktop)</summary>

```
Restarted application in 534ms.
flutter: elapsed 6306ms LogicalKeySet#00106(keys: Meta Left)
flutter: elapsed 6513ms LogicalKeySet#e485d(keys: Meta Left + Key V)
flutter: elapsed 6666ms LogicalKeySet#e485d(keys: Meta Left + Key V)
flutter: elapsed 6806ms LogicalKeySet#e485d(keys: Meta Left + Key V)
flutter: elapsed 6915ms LogicalKeySet#e485d(keys: Meta Left + Key V)
flutter: elapsed 7051ms LogicalKeySet#e485d(keys: Meta Left + Key V)
flutter: elapsed 7212ms LogicalKeySet#e485d(keys: Meta Left + Key V)
flutter: elapsed 7334ms LogicalKeySet#e485d(keys: Meta Left + Key V)
flutter: elapsed 7506ms LogicalKeySet#e485d(keys: Meta Left + Key V)
flutter: elapsed 7671ms LogicalKeySet#e485d(keys: Meta Left + Key V)
flutter: elapsed 7852ms LogicalKeySet#e485d(keys: Meta Left + Key V)
flutter: elapsed 8015ms LogicalKeySet#e485d(keys: Meta Left + Key V)
flutter: elapsed 8139ms LogicalKeySet#e485d(keys: Meta Left + Key V)
flutter: elapsed 8319ms LogicalKeySet#e485d(keys: Meta Left + Key V)
flutter: elapsed 8507ms LogicalKeySet#e485d(keys: Meta Left + Key V)
flutter: elapsed 8627ms LogicalKeySet#e485d(keys: Meta Left + Key V)
flutter: elapsed 8756ms LogicalKeySet#e485d(keys: Meta Left + Key V)
flutter: elapsed 8859ms LogicalKeySet#e485d(keys: Meta Left + Key V)
flutter: elapsed 9047ms LogicalKeySet#e485d(keys: Meta Left + Key V)
flutter: elapsed 9127ms LogicalKeySet#e485d(keys: Meta Left + Key V)
flutter: elapsed 9301ms LogicalKeySet#e485d(keys: Meta Left + Key V)
flutter: elapsed 9370ms LogicalKeySet#e485d(keys: Meta Left + Key V)
flutter: elapsed 9570ms LogicalKeySet#e485d(keys: Meta Left + Key V)
flutter: elapsed 9670ms LogicalKeySet#e485d(keys: Meta Left + Key V)
flutter: elapsed 9740ms LogicalKeySet#e485d(keys: Meta Left + Key V)
flutter: elapsed 9906ms LogicalKeySet#e485d(keys: Meta Left + Key V)
flutter: elapsed 10095ms LogicalKeySet#e485d(keys: Meta Left + Key V)
flutter: elapsed 10270ms LogicalKeySet#e485d(keys: Meta Left + Key V)
flutter: elapsed 10373ms LogicalKeySet#e485d(keys: Meta Left + Key V)
flutter: elapsed 10492ms LogicalKeySet#e485d(keys: Meta Left + Key V)
flutter: elapsed 10671ms LogicalKeySet#e485d(keys: Meta Left + Key V)
flutter: elapsed 10826ms LogicalKeySet#e485d(keys: Meta Left + Key V)
flutter: elapsed 10963ms LogicalKeySet#e485d(keys: Meta Left + Key V)
flutter: elapsed 11138ms LogicalKeySet#e485d(keys: Meta Left + Key V)
flutter: elapsed 11294ms LogicalKeySet#e485d(keys: Meta Left + Key V)
flutter: elapsed 11424ms LogicalKeySet#e485d(keys: Meta Left + Key V)
flutter: elapsed 11556ms LogicalKeySet#e485d(keys: Meta Left + Key V)
flutter: elapsed 11742ms LogicalKeySet#e485d(keys: Meta Left + Key V)
flutter: elapsed 11862ms LogicalKeySet#e485d(keys: Meta Left + Key V)

```
</details>


<details>
<summary>flutter doctor -v</summary>

```
[✓] Flutter (Channel stable, 2.8.1, on macOS 12.1 21C52 darwin-arm, locale en-GB)
    • Flutter version 2.8.1 at /Users/mahesh/Documents/flutter
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision 77d935af4d (6 days ago), 2021-12-16 08:37:33 -0800
    • Engine revision 890a5fca2e
    • Dart version 2.15.1

[✓] Android toolchain - develop for Android devices (Android SDK version 31.0.0)
    • Android SDK at /Users/mahesh/Library/Android/sdk
    • Platform android-31, build-tools 31.0.0
    • ANDROID_HOME = /Users/mahesh/Library/Android/sdk
    • Java binary at: /Applications/Android Studio.app/Contents/jre/jdk/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 11.0.8+10-b944.6916264)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS (Xcode 13.1)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • CocoaPods version 1.10.2

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] Android Studio (version 4.2)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 11.0.8+10-b944.6916264)

[✓] IntelliJ IDEA Community Edition (version 2021.2.1)
    • IntelliJ at /Applications/IntelliJ IDEA CE.app
    • Flutter plugin version 60.1.4
    • Dart plugin version 212.5080.8

[✓] VS Code (version 1.61.2)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.29.0

[✓] Connected device (3 available)
    • iPhone 8 (mobile) • A55680BB-E46F-425E-B803-ACB8DEA8A239 • ios            •
      com.apple.CoreSimulator.SimRuntime.iOS-15-0 (simulator)
    • macOS (desktop)   • macos                                • darwin-arm64   • macOS 12.1
      21C52 darwin-arm
    • Chrome (web)      • chrome                               • web-javascript • Google
      Chrome 96.0.4664.110

• No issues found!
```

```
[✓] Flutter (Channel master, 2.9.0-1.0.pre.93, on macOS 12.1 21C52 darwin-arm, locale en-GB)
    • Flutter version 2.9.0-1.0.pre.93 at /Users/mahesh/Documents/flutter_master
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision 1be7f1ea56 (7 days ago), 2021-12-14 22:14:07 -0500
    • Engine revision e444009bf4
    • Dart version 2.16.0 (build 2.16.0-109.0.dev)
    • DevTools version 2.9.1

[✓] Android toolchain - develop for Android devices (Android SDK version 31.0.0)
    • Android SDK at /Users/mahesh/Library/Android/sdk
    • Platform android-31, build-tools 31.0.0
    • ANDROID_HOME = /Users/mahesh/Library/Android/sdk
    • Java binary at: /Applications/Android Studio.app/Contents/jre/jdk/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 11.0.8+10-b944.6916264)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS (Xcode 13.1)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • CocoaPods version 1.10.2

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] Android Studio (version 4.2)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 11.0.8+10-b944.6916264)

[✓] IntelliJ IDEA Community Edition (version 2021.2.1)
    • IntelliJ at /Applications/IntelliJ IDEA CE.app
    • Flutter plugin version 60.1.4
    • Dart plugin version 212.5080.8

[✓] VS Code (version 1.61.2)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.29.0

[✓] Connected device (3 available)
    • iPhone 8 (mobile) • A55680BB-E46F-425E-B803-ACB8DEA8A239 • ios            •
      com.apple.CoreSimulator.SimRuntime.iOS-15-0 (simulator)
    • macOS (desktop)   • macos                                • darwin-arm64   • macOS 12.1
      21C52 darwin-arm
    • Chrome (web)      • chrome                               • web-javascript • Google
      Chrome 96.0.4664.110

• No issues found!
```
</details>

**Comment by mdebbar** [comment]
Thanks for the detailed explanation @luigi-rosso!

The person who's most familiar with our current keyboard implementation is @dkwingsmt so I would like to get his opinion on this. 

> Our workaround takes a different approach, we will always report keydowns when modifiers are pressed and instead of synthesizing on a timer, we synthesize key-ups for all known pressed keys when the modifier is finally released. Rive's fix will report:
> 
> 0ms: Press CMD 200ms: Press V 600ms: Press V (we could also synthesize a key-up right before this, but we currently don't). 800ms: Release V (sent just before CMD as we see V was pressed while CMD was down, so the release of CMD releases all other pressed keys in context of it) 800ms: Release CMD

@dkwingsmt do you think this approach makes sense for us? Do you foresee any issues with it?
@luigi-rosso how do you handle cases that involve multiple modifier keys?

**My 2 cents:**
Given that this is fundamentally a browser issue, I don't think we can have a perfect solution that works in all cases. So we either have to make a trade-off, or let the user make the trade-off by giving them the option to choose which approach to use (our current approach vs @luigi-rosso's workaround approach).

**Comment by dkwingsmt** [comment]
@luigi-rosso Thank you very much for the detailed explanation!

I agree that it's a concerning problem and your solution makes a lot of sense. Though, I have some questions: Are you using the `HardwareKeyboard`/`KeyEvent` system, or the `RawKeyboard`/`RawKeyEvent` system? Currently the `Shortcut` class in the Flutter framework is still based on `RawKeyEvent`, and the situation you described should not happen (and indeed not according to my test.)

So I'll assume you're using the newer `KeyEvent` (which I appreciate.) Now this explains because its regularization algorithm is designed so that back-to-back key down events are resolved by ignoring the second event, which is described [here](https://github.com/flutter/engine/blob/9177b32b75a8dfa3deb4f8823003ac219cea7c13/lib/web_ui/lib/src/engine/keyboard_binding.dart#L413). To be honest, this decision was made pretty arbitrarily, so I'm willing to change it, and I think this should be a more universal solution than yours.

To test, can you change L413 and 414 in the file above to:

```diff
-          event.preventDefault();
-          return;
+          type = ui.KeyEventType.repeat;
```

This change should make the second event a KeyRepeatEvent instead. Can you try and see if your problem is resolved? (I'd like to test but I'm still building.)

Another option is to simulate a KeyUpEvent before the second down event, similar to your plan. Do you have a preference over these two?

**Comment by luigi-rosso** [comment]
Hi @dkwingsmt! I was actually using the old RawKeyboard system, but not with Shortcuts at the time. I was just listening to the onKey callback of a FocusNode.  I was excited to see the new HardwareKeyboard changes get merged (thank you!) but I tried there too (using onKeyEvent) and it seems like the gating logic is very similar if not the same.

Oh your solution is interesting, and it would solve this problem but it would exacerbate another problem. In a separate issue us we're trying to filter synthesized repeats: https://github.com/flutter/flutter/issues/95506

Basically, we only want certain actions to happen on actual hardware key-presses, so sending through a repeat wouldn't work for us as we'd try to filter that out at the app layer. So masquerading this second press as a repeat would make us need to not filter the repeats (which breaks the requirements on the other issue).

**Comment by dkwingsmt** [comment]
That's really weird. Can you isolate an example to demonstrate the problem? The problem does not seem to occur when using `Shortcut`.

`keyboard_binding.dart` only changes the `HardwareKeyboard` system.

Yes, `HardwareKeyboard` is designed to be as seamless as possible (while solving some very edge cases that had been bugging us.)

**Comment by luigi-rosso** [comment]
@mdebbar we constantly sync all modifier key-presses (the js event always carries whether any mod key is down) but currently only release all other pressed keys on meta release.  It works for the shortcuts we have in place, but is not a great generic solution.

Here's the rundown of how we do it from JS land:
```js
    // initInputHelper is called by our Dart code during app init. keyPressed is
    // also a function defined in Dart that we call from JS to send the key code
    // with some bit flags to indicate it is a press/release and whether it's a
    // repeat. That will (in Dart) send it via our custom MethodChannel.
    function initInputHelper() {

      const pressedKeys = new Set();

      const metaKeyCode = 0x5B;
      const ctrlKeyCode = 0x11;
      const shiftKeyCode = 0x10;
      const altKeyCode = 0x12;

      function isModifier(code) {
        switch (code) {
          case metaKeyCode:
          case ctrlKeyCode:
          case shiftKeyCode:
          case altKeyCode:
            return true;
          default:
            return false;
        }
      }

      window.addEventListener("focus", function (event) {
        // Sync modifiers to released state.
        reportKey(metaKeyCode, false);
        reportKey(ctrlKeyCode, false);
        reportKey(shiftKeyCode, false);
        reportKey(altKeyCode, false);
      });

      function reportKey(code, isPress, isRepeat) {
        if (isModifier(code) && isPress && pressedKeys.has(code)) {
          // We already think this modifier key is pressed. We allow reporting
          // other keys multiple times.
          return false;
        }
        if (!isPress && !pressedKeys.has(code)) {
          // We already think the key is not pressed.
          return false;
        }

        // We need to report the key.
        if (isPress) {
          pressedKeys.add(code);
        } else {
          pressedKeys.delete(code);
        }

        if (!isPress) {
          code |= 1 << 18;
        }
        if (isRepeat) {
          code |= 1 << 17;
        }
        keyPressed(code);
        if (maskedKeys.has(code)) {
          return false;
        }
        return true;
      }

      function syncModifiers(event) {
        if (reportKey(metaKeyCode, event.metaKey) && !event.metaKey) {
          // If meta was released, we need to notify that all other non modifier
          // keys were released too.
          for (const code of pressedKeys) {
            if (!isModifier(code) && code != event.keyCode) {
              reportKey(code, false);
            }
          }
        }

        // Note that we also take this opportunity to sync modifiers that may
        // have not reported keydown/up but will report changed in the event
        // datastructure. reportKey will do the gating for us if necessary.
        reportKey(ctrlKeyCode, event.ctrlKey);
        reportKey(shiftKeyCode, event.shiftKey);
        reportKey(altKeyCode, event.altKey);
      }

      window.addEventListener("keydown", function (event) {
        syncModifiers(event);
        const propagate = reportKey(event.keyCode, true, event.repeat);
        if (!propagate) {
          event.stopPropagation();
          event.preventDefault();
        }
      });

      window.addEventListener("keyup", function (event) {
        syncModifiers(event);
        reportKey(event.keyCode, false);
      });
    }
```

**Comment by luigi-rosso** [comment]
> That's really weird. Can you isolate an example to demonstrate the problem? The problem does not seem to occur when using `Shortcut`.
> 
> `keyboard_binding.dart` only changes the `HardwareKeyboard` system.
> 
> Yes, `HardwareKeyboard` is designed to be as seamless as possible (while solving some very edge cases that had been bugging us.)

Let me try to spin up an example! Would be great if it worked there!

**Comment by luigi-rosso** [comment]
@dkwingsmt you are right! Meta + V, V, V, V reports multiple Paste actions with Shortcuts/old keyboard api. However, it breaks when interjecting a different key:

For example, when trying to copy-paste, a common practice is to keep meta held while pressing C and V in succession:

Press Meta
Press C (copy action occurs)
Release C (keep Meta pressed)
Press V (paste action should occur, does not until 2s gate is opened)

I'll get you a demo that makes this clear!

**Comment by luigi-rosso** [comment]
Here's an example that shows the issue.  You'll see in the recording that I'm pressing Meta + C, then keeping Meta pressed I release C and press V. My mac reports the Meta V, but the Paste action is never fired in Flutter Web. Eventually when I press it enough times and the 2s gate passes, it works but in the meantime the user feels like their paste is broken.

![2021-12-22 17 12 09](https://user-images.githubusercontent.com/454182/147172794-058d8ed1-98aa-4a24-8d87-3105828d3c29.gif)

Code here: https://gist.github.com/luigi-rosso/08009f82ec10f0581a2fd204ce03ee9d

**Comment by dkwingsmt** [comment]
@luigi-rosso Thanks for writing this example! 

First off, I managed to reproduce this error on my own and fixed it. For the record, this is what I did:

App: A normal `TextField`.

Steps: 
1. Run with `flutter run -d chrome`.
2. Type a few characters in the text field.
3. Hold Cmd, then rapidly press Left and Right alternately.

**Comment by dkwingsmt** [comment]
However, there are more problems in your code.

There's a minor issue: On L117 and L124, `CopyIntent` should be changed to `PasteIntent`.

And there's a major issue: On L18 and L20, `LogicalKeySet` must be migrated to `SingleActivator`. `LogicalKeySet` is the legacy class. In most cases, migrating it to `SingleActivator` is optional, but in this case it's mandatory. It's because `LogicalKeySet` uses an exact match, while `SingleActivator` allows a "more vague" kind of match.

Say you want to write a shortcut for key combination `Meta + C`. If you write it using `LogicalKeySet`, then the key combination will be recognized when "any key down events occur, and the current pressed keys are exactly {Meta, KeyC}". (Let's assume there's only one Meta key, since it takes care of MetaLeft and MetaRight.) The problem is, since it is impossible to detect when the previous `Meta  + V` is released on Web, therefore it is assumed that both Meta and KeyV are being held until the end of the "key guard window", when you press `C` later, the current pressed keys are actually `{Meta, KeyC, KeyV}`. And `LogicalKeySet` rejects it, until `KeyV` is not pressed for 2 seconds.

What's worse, due to the same reason, if you press `Meta + C`, release both keys, then rapidly press `Meta`, guess what: the shortcut is recognized for each `Meta` press for 2 seconds, even when `KeyC` is no longer held! (Of course, it also means that, pressing Meta while holding KeyC always triggers the shortcut, which is why we decided to abandon the class in the first place.)

On the other hand, `SingleActivator` uses a different policy. It only checks when the currently event is "the non-modifier key", and performs an exact match of the modifier keys. In the example above, it finds all key combinations with a `KeyC`, and sees if any of them requires "pressing Meta, and none of the other modifier keys". So even when `KeyV` is incorrectly considered as held at that time, it doesn't prevent `Meta + C` being recognized! 

-----

So, in short, please kindly wait until the PR is merged and rolled, and migrate all occurrences of `LogicalKeySet` in your codebase to `SingleActivator`.

**Comment by luigi-rosso** [comment]
Thanks @dkwingsmt! How can I see when a Flutter channel has rolled to an engine with this PR? Should it already be in Flutter master?

**Comment by HansMuller** [comment]
The `cp: 2.10` label is for https://github.com/flutter/engine/pull/30488, which will hopefully cause that PR to be included in the next stable Flutter release as well as the current Flutter master branch.

**Comment by dkwingsmt** [comment]
@luigi-rosso You can find a guide [here](https://github.com/flutter/flutter/wiki/Where%27s-my-Commit%3F#finding-the-framework-commit-that-contains-engine-commit-x). At the time of writing, this engine commit has not been rolled into the framework yet. We have decided to (propose to) cherrypick it into the next stable release, 2.10.

**Comment by luigi-rosso** [comment]
Thank you!

**Comment by luigi-rosso** [comment]
I just tried this in master and can confirm it's working as expected on both desktop and web! :tada: Thanks @dkwingsmt et al!
